lwt-unix/httpaf_lwt_unix.ml

```diff
@@ -0,0 +1,128 @@
+(*----------------------------------------------------------------------------
+    Copyright (c) 2018 Inhabited Type LLC.
+
+    All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions
+    are met:
+
+    1. Redistributions of source code must retain the above copyright
+       notice, this list of conditions and the following disclaimer.
+
+    2. Redistributions in binary form must reproduce the above copyright
+       notice, this list of conditions and the following disclaimer in the
+       documentation and/or other materials provided with the distribution.
+
+    3. Neither the name of the author nor the names of his contributors
+       may be used to endorse or promote products derived from this software
+       without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE CONTRIBUTORS ``AS IS'' AND ANY EXPRESS
+    OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+    DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
+    ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+    OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+    HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+    STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+  ----------------------------------------------------------------------------*)
+
+open Lwt.Infix
+open Httpaf
+
+module Server = struct
+  let create_connection_handler ?config ~request_handler ~error_handler =
+    fun client_addr sock ->
+
+    let conn =
+      Server_connection.create ?config
+        ~error_handler:(error_handler client_addr)
+        (request_handler client_addr) in
+
+    let buffer = Lwt_bytes.create 0x1000 in (* TODO: Make configurable. *)
+    let rec reader_thread () =
+      match Server_connection.next_read_operation conn with
+      | `Read ->
+        Lwt_bytes.read sock buffer 0 (Lwt_bytes.length buffer) >>= fun len ->
+        (if len = 0
+         then Server_connection.shutdown_reader conn
+         else Server_connection.read conn buffer ~off:0 ~len |> ignore);
```

Comment:
The server connection might not consume the entire buffer here. For example, if the input does not end on a token boundary. The code needs to check the return value of `read` and hold onto any bytes that were not consumed  for the next call to `read`.